crates/fj-core/src/validate/sketch.rs

```diff
@@ -1,12 +1,117 @@
-use crate::objects::Sketch;
+use crate::{
+    objects::{HalfEdge, Sketch},
+    storage::Handle,
+};
 
 use super::{Validate, ValidationConfig, ValidationError};
 
 impl Validate for Sketch {
     fn validate_with_config(
         &self,
-        _: &ValidationConfig,
-        _: &mut Vec<ValidationError>,
+        config: &ValidationConfig,
+        errors: &mut Vec<ValidationError>,
     ) {
+        SketchValidationError::check_obect_references(self, config, errors);
+    }
+}
+
+/// [`Sketch`] validation failed
+#[derive(Clone, Debug, thiserror::Error)]
+pub enum SketchValidationError {
+    /// [`HalfEdge`] referenced by more than one [`Cycle`]
+    #[error("[`HalfEdge`] referenced by more than one [`Cycle`]")]
+    HalfEdgeMultipleReferences,
+    /// [`Cycle`] referenced by more than one [`Region`]
+    #[error("[`Cycle`] referenced by more than one [`Region`]")]
+    CycleMultipleReferences,
+}
+
+impl SketchValidationError {
+    fn check_obect_references(
+        sketch: &Sketch,
+        _config: &ValidationConfig,
+        errors: &mut Vec<ValidationError>,
+    ) {
+        let mut referenced_edges =
+            std::collections::HashMap::<Handle<HalfEdge>, i32>::new();
+
+        // Do we care about how many times each edge is used, or should we just return as soon as
+        // we find one that is used more than once?
```

Comment:
Experience has shown that early-aborting validation checks is generally a bad idea (which is why they're done in the background through a service, instead of using the normal Rust `Result`-based error handling).

With early abort, you often end up with not enough information to figure out what's going wrong.